CONVENTIONS.md

```diff
@@ -0,0 +1,118 @@
+# Naming conventions
+
+## General
+
+Use `UpperCamelCase` for class names. Use `lowerCamelCase` for method names, variable names, and names of fields that are not
+static and final. Always use `UPPER_SNAKE_CASE` for names of fields that are both static and final.
+
+Method names should always be verbal phrases (`tick`, `getCarversForStep`). Non-boolean field and variable names and class names
+should be noun phrases (`ChunkRegion`, `color`). Boolean field and variable names should always be adjectival phrases or present
+tense verbal phrases (`powered`, `canOpen`), avoiding the `is` and `has` prefixes when possible (`colored`, not `isColored` or
+`hasColor`).
+
+To make code as easy to read as possible, keep names in the natural language order. For example, a class representing a "chest
+block entity" should be named `ChestBlockEntity`, not `BlockEntityChest`. Though prefix naming may be helpful for grouping
+classes together in an IDE's tree view, reading and writing code is done much more often than browsing files.
+
+Use American English (`color`, not `colour`) for consistency throughout Yarn and with known Mojang names.
+
+Omit words that are redundant because of method parameter or the owner class's name. For example, use `getChunk(BlockPos pos)`
+rather than `getChunkAtPosition(BlockPos pos)`, and `create` rather than `createAchievement` for a method is in the
+`Achievement` class. Don't avoid overloading methods or shadowing fields.
+
+Names should be as descriptive as possible, so don't omit important words. When naming something always look at all its usages
+(including overriding methods and inheriting classes).
+
+Consistency is important as it makes code more readable and names easier to memorize. When possible, use terms that are present
+in other Yarn names, in libraries used by Minecraft, or in vanilla strings. See the "Common names" and "Mojang names" for more
+information.
+
+## Acronym and abbreviations
+
+Try to avoid acronyms and abbreviations unless it's a common one that everyone knows, and other yarn names involving the same
+word all use its abbreviated form. Full names are easier to read quickly and remember ("Which words were abbreviated?") and
+they often don't take more time to type thanks to IDE autocompletion. The common abbreviations you *should* use are:
+
+ - "id" for "identifier"
+ - "pos" for "position"
+ - "NBT" for "named binary tag"
+ - "init" for "initialize"
+ - Any abbreviations used by Java or libraries ("json", "html", etc.)
+
+Treat acronyms as single words rather than capitalising every letter. This improves readability (compare `JsonObject` and
+`JSONObject`) and is consistent with Mojang naming (a known name is `NbtIo`).
+
+## Mojang names
+
+Always use names that match names in strings in the vanilla code, unless that string is outdated or inaccurate. This avoids
+confusion, especially from new modders who may not understand what an class exception message is referring to.
```

Comment:
Nah, mojang names are usually confusing and should be considered separately. e.g. EntitySize was added before mojang added a toString that says it is called `EntityDimension`, which is confusing since `Dimension` is already present.

Reply:
@liach `EntityDimensions` is a much better name than than entity size. As I [explained in a comment on that issue](https://github.com/FabricMC/yarn/issues/715#issuecomment-498521419), size and dimension don't mean the same thing: a size is a single value telling you how big something is overall, while the dimensions are how big something is in each direction.

In any case, the point of this section isn't to always use Mojang names for everything, but rather to use terms that are more likely to be understood by everyone. For example, if Mojang calls something an "entity goal", don't call it an "ai target", or players will get confused when they get an exception message mentioning goals.

I'll edit this section and make this more clear, and mention that names should also match the terms commonly used by Minecraft players and modders.